Bump to 0.28.0

ChristofKaufmann opened the pull request:
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

@pseudo-rnd-thoughts I see you as contributor in [farama-notifications](https://github.com/Farama-Foundation/Farama-Notifications). Do you want to be maintainer for the [farama-notifications-feedstock](https://github.com/conda-forge/farama-notifications-feedstock)? I added the package on conda-forge for this version update.

@h-vetinari Do you want to be added as well?

h-vetinari commented:
Thanks for the PR, and taking care of the new dependency!

> @h-vetinari Do you want to be added as well?

I think it's in good hands with you (and possibly @pseudo-rnd-thoughts). :)

Also, I think we can skip 0.28.0 here and go for 0.28.1 directly.

h-vetinari commented:
Also, I seem to remember (perhaps incorrectly) that starting from 0.28, we can drop the dependency on `gym`?

ChristofKaufmann commented:
> Also, I think we can skip 0.28.0 here and go for 0.28.1 directly.

Yes, that would have been a similar amount of work. I misinterpreted the comment [here](https://github.com/conda-forge/gymnasium-feedstock/pull/16#issuecomment-1548173011). However, going from 0.28.0 to 0.28.1 will be easier. It seems only patch 0003 needs to be updated.

> Also, I seem to remember (perhaps incorrectly) that starting from 0.28, we can drop the dependency on gym?

I don't know of any change regarding gym between 0.27 and 0.28. So maybe we could have dropped it before. Should I remove it from the test dependencies in this PR? IMO keeping it does not really hurt, but increase test coverage. Maybe upstream gymnasium should drop the support, if it is not tested. I think the [documentation](https://gymnasium.farama.org/content/gym_compatibility/) about it is outdated as well. But this discussion does not really belong here.

h-vetinari commented:
> > Also, I think we can skip 0.28.0 here and go for 0.28.1 directly.
> 
> Yes, that would have been a similar amount of work. I misinterpreted the comment [here](https://github.com/conda-forge/gymnasium-feedstock/pull/16#issuecomment-1548173011). However, going from 0.28.0 to 0.28.1 will be easier. It seems only patch 0003 needs to be updated.

You should still go to 0.28.1 in this PR.

ChristofKaufmann commented:
> We need to reflect the [dependencies](https://github.com/Farama-Foundation/Gymnasium/blob/v0.28.1/pyproject.toml#L27-L84) as defined upstream.

That is not the case yet. We would need to make some changes:

- typing_extensions → typing-extensions >=4.3.0
- importlib_metadata >=4.8.1 → importlib_metadata >=4.8.0
- pytest → pytest ==7.1.3 (only for testing)
- scipy → scipy ==1.7.3 (only for testing)
- shimmy-atari → shimmy-atari >=0.1.0,<1.0 (although we tested with 1.0.0 successfully)
- pygame ==2.1.* → pygame ==2.1.3 (not really a difference, since 2.1.3 is the only version matching 2.1.3)
- box2d-py ==2.3.* → box2d-py ==2.3.5 (cannot be done, since only 2.3.8 is on conda-forge)
- mujoco-python >=2.2.0,<=2.3.3 → mujoco-python <=2.3.3 (was >=2.3.2 in upstream 0.28.1)
- ale-py >=0.8.0,<0.9 should be removed from gymnasium-atari
- pytorch >=1.0.0 is missing in gymnasium-other
- gymnasium-jax with dependencies "jax ==0.3.24", "jaxlib ==0.3.24" is missing completely.

Maybe in another PR to get 0.28.1 pushed now?

h-vetinari commented:
> Maybe in another PR to get 0.28.1 pushed now?

No. We don't push with knowingly ~wrong~ suboptimal dependency information. Some of the things you describe are valid replacements, some need fixing.
> typing_extensions → typing-extensions >=4.3.0

the underscore vs. hyphen is irrelevant, conda-forge supports both spellings. However, I don't see this requirement upstream at all, why is it needed?

> importlib_metadata >=4.8.1 → importlib_metadata >=4.8.0

irrelevant difference, but similarly, I don't see this upstream

> pytest → pytest ==7.1.3 (only for testing)

We never pin pytest unless absolutely necessary

> scipy → scipy ==1.7.3 (only for testing)

Don't add pin for testing. Unless we add a `run_constrained:` to the respective output, users will be able to co-install this with the latest scipy, and so we need to test that

> shimmy-atari → shimmy-atari >=0.1.0,<1.0 (although we tested with 1.0.0 successfully)

Judgement call. Normally we should follow upstream pins (assuming there are good reasons for them), but if the tests pass it should be alright.

> pygame ==2.1.* → pygame ==2.1.3 (not really a difference, since 2.1.3 is the only version matching 2.1.3)

Fine to leave as-is

> box2d-py ==2.3.* → box2d-py ==2.3.5 (cannot be done, since only 2.3.8 is on conda-forge)

Fine to leave as-is

> mujoco-python >=2.2.0,<=2.3.3 → mujoco-python <=2.3.3 (was >=2.3.2 in upstream 0.28.1)

Change to `>=2.3.2` as upstream

> ale-py >=0.8.0,<0.9 should be removed from gymnasium-atari

Remove it. Sidenote: ain't I glad we spent so much time bringing this to conda-forge specifically for gym/gymnasium.

> pytorch >=1.0.0 is missing in gymnasium-other

Add pytorch to gymnasium-other, remove it from gymnasium-all

> gymnasium-jax with dependencies "jax ==0.3.24", "jaxlib ==0.3.24" is missing completely.

That's the only thing we can (and probably should) defer to a separate PR

ChristofKaufmann commented:
> > typing_extensions → typing-extensions >=4.3.0
>
> the underscore vs. hyphen is irrelevant, conda-forge supports both spellings. However, I don't see this requirement upstream at all, why is it needed?

It is right [here](https://github.com/Farama-Foundation/Gymnasium/blob/v0.28.1/pyproject.toml#L32), where I also copied the hyphen from.

Changed to `typing_extensions >=4.3.0`.

> > importlib_metadata >=4.8.1 → importlib_metadata >=4.8.0
>
> irrelevant difference, but similarly, I don't see this upstream

It is [here](https://github.com/Farama-Foundation/Gymnasium/blob/v0.28.1/pyproject.toml#L31).

Left as it is: `importlib_metadata >=4.8.1  # [py<=39]`

> > pytest → pytest ==7.1.3 (only for testing)
>
> We never pin pytest unless absolutely necessary
>
> > scipy → scipy ==1.7.3 (only for testing)
>
> Don't add pin for testing. Unless we add a `run_constrained:` to the respective output, users will be able to co-install this with the latest scipy, and so we need to test that

I am ok with that, but these pinnings would only apply during the test in the build process.

Left as it is: `scipy` and `pytest`.

> > shimmy-atari → shimmy-atari >=0.1.0,<1.0 (although we tested with 1.0.0 successfully)
>
> Judgement call. Normally we should follow upstream pins (assuming there are good reasons for them), but if the tests pass it should be alright.

Left as it is: `shimmy-atari`

> > pygame ==2.1.* → pygame ==2.1.3 (not really a difference, since 2.1.3 is the only version matching 2.1.3)
>
> Fine to leave as-is
>
> > box2d-py ==2.3.* → box2d-py ==2.3.5 (cannot be done, since only 2.3.8 is on conda-forge)
>
> Fine to leave as-is

Left as it is: `box2d-py ==2.3.*` and `pygame ==2.1.*`

> > mujoco-python >=2.2.0,<=2.3.3 → mujoco-python <=2.3.3 (was >=2.3.2 in upstream 0.28.1)
>
> Change to `>=2.3.2` as upstream

Changed to `mujoco-python >=2.3.2,<=2.3.3`, because 2.3.4 is incompatible, see [this](https://github.com/deepmind/mujoco/issues/833) MuJoCo issue and this Gymnasium [PR](https://github.com/Farama-Foundation/Gymnasium/pull/463).

> > ale-py >=0.8.0,<0.9 should be removed from gymnasium-atari
>
> Remove it. Sidenote: ain't I glad we spent so much time bringing this to conda-forge specifically for gym/gymnasium.

Removed. Note, it is now an indirect dependency: `gymnasium-atari`→ `shimmy-atari` → `ale-py >=0.8.1,<0.9`. It is still required for the Atari environments, but shimmy provides the interface.

> > pytorch >=1.0.0 is missing in gymnasium-other
>
> Add pytorch to gymnasium-other, remove it from gymnasium-all

Moved.

h-vetinari commented:
OK, thanks for the updates. Not sure how I managed to overlook the typing_extensions, I even did CTRL+F in that file. 🤷 

> I am ok with that, but these pinnings would only apply during the test in the build process.

You might mean the right thing, but for clarity: the CI job here is separated into a build & a test phase, which are very much insulated from each other. The only thing that passes through is the newly-installed artefact. So there's no testing during the build process[^1], and it wouldn't make sense either, because it's not what we want to test (i.e. the installed artefact rather than files in the `$SRC_DIR`). And on top of that, running CI here with a pinned scipy or pytest version just means we might miss issues with the most common installation pattern, which is with the newest versions of both.

[^1]: at least for Python packages, C/C++ libraries are a different story

h-vetinari reviewed (approved):
Thanks a lot for this @ChristofKaufmann! :)

I'd still like @pseudo-rnd-thoughts to have a look, but this LGTM!

pseudo-rnd-thoughts commented:
Amazing, thank you @ChristofKaufmann. Looks good to me
